### PR TITLE
add resetting histogram type and resetting counter type

### PR DESCRIPTION
**Pull request by nonsense**
This PR is adding:
1. a `ResettingHistogram` - one that we can use for measuring latencies.
2. a `ResettingCounter` - one that gets reset every time we emit it to InfluxDB, so that we can count easily on Grafana side (See screenshots and comments).

TODO:
- [x] implement `Count`
- [x] implement `Min`
- [x] implement `Max`
- [x] implement `Variance`
- [x] implement `Sum`
- [x] implement `StdDev`
- [x] implement `Clear`

for all 3 types - `Nil`, `Snapshot` and `StandardResettingHistogram`.

### CODE REVIEW

File: runtime/metrics_api.go [hunk 1]
```diff
@@ -172,6 +172,10 @@ func (m *MetricsApi) Timer(name string) Timer {
 	return m.reg.GetOrRegister(name, metrics.NewTimer()).(metrics.Timer)
 }
 
+func (m *MetricsApi) ResettingTimer(name string) ResettingTimer {
+	return m.reg.GetOrRegister(name, NewResettingTimer()).(ResettingTimer)
```
Comment (nonsense) [inline runtime/metrics_api.go]: @raulk I am not sure how we are going to extend the go-metrics library to support new types, because:

```
func (r *StandardRegistry) register(name string, i interface{}) error {
	if _, ok := r.metrics[name]; ok {
		return DuplicateMetric(name)
	}
	switch i.(type) {
	case Counter, Gauge, GaugeFloat64, Healthcheck, Histogram, Meter, Timer:
		r.metrics[name] = i
	}
	return nil
}
```

Comment (raulk) [reply runtime/metrics_api.go]: @nonsense 

`Counter, Gauge, GaugeFloat64, Healthcheck, Histogram, Meter, Timer` <= these are interfaces, right? As long our new structs implement those interfaces, we should be good to go?

Comment (nonsense) [reply runtime/metrics_api.go]: As discussed, renaming `ResettingTimer` to `ResettingHistogram` and making sure it implements the `Histogram` interface.

File: runtime/metrics_types.go [hunk 2]
```diff
@@ -111,19 +112,16 @@ func NewMetric(name string, i interface{}) *Metric {
 		t = MetricHistogram
 		m = pools[t].Get().(*Metric)
 		s := v.Snapshot()
-		p := s.Percentiles([]float64{0.5, 0.75, 0.95, 0.99, 0.999, 0.9999})
+		p := s.Percentiles([]float64{0.5, 0.75, 0.95, 0.99, 0.999})
 		m.Measures["count"] = float64(s.Count())
 		m.Measures["max"] = float64(s.Max())
 		m.Measures["mean"] = s.Mean()
 		m.Measures["min"] = float64(s.Min())
-		m.Measures["stddev"] = s.StdDev()
```
Comment (nonsense) [inline runtime/metrics_types.go]: @raulk I've add calculations for `stddev` and `variance`, but I really don't think we need to be calculating these at runtime and sending them to InfluxDB.

If for some reason someone needs them, we can always enable them. WDYT?

File: runtime/metrics_api.go [hunk 3]
```diff
@@ -145,7 +145,7 @@ func (m *MetricsApi) RecordPoint(name string, value float64) {
 }
 
 func (m *MetricsApi) Counter(name string) Counter {
-	return m.reg.GetOrRegister(name, metrics.NewCounter()).(metrics.Counter)
+	return m.reg.GetOrRegister(name, NewResettingCounter()).(metrics.Counter)
```
Comment (nonsense) [inline runtime/metrics_api.go]: @raulk I think by default it makes it easier to use the resetting counter, than the counter. It resets every time we emit it (i.e. when calling `Snapshot`).

The difference is the following:

Resetting counter:
Screenshot 2020-05-13 at 15 28 09

Default counter:
Screenshot 2020-05-13 at 15 29 00

The counter you are looking at is 50 instances increasing a counter 5 times (50 instances starting 5 listeners) == total 250 `listens.ok`.

The second screenshot is wrong, the sum total value has to be 250, but since this is an accumulating counter, we need to add some kind of query to subtract adjacent values. It is much easier to use a statsd counter type (rather than prometheus type) and just reset it when emitting it. Obviously with the default counter we can't use the `sum` aggregation in InfluxDB/Grafana (hence the wrong results).

File: runtime/metrics.go [hunk 4]
```diff
@@ -47,7 +47,7 @@ func newMetrics(re *RunEnv) *Metrics {
 	}
 
 	m.diagnostics = newMetricsApi(re, metricsApiOpts{
-		freq:        5 * time.Second,
```
Comment (nonsense) [inline runtime/metrics.go]: I think 5 seconds is a bit aggressive, especially when we want to run 10k instances. We should probably think about making this configurable, but for now 10sec. or 30sec. is good enough.

File: runtime/resetting_counter.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+// Inc increments the counter by the given amount.
+func (c *StandardResettingCounter) Inc(i int64) {
+	atomic.AddInt64(&c.count, i)
+}
+
+// Snapshot returns a read-only copy of the counter, and resets it.
+func (c *StandardResettingCounter) Snapshot() Counter {
+	currentValue := atomic.SwapInt64(&c.count, 0)
```
Comment (nonsense) [inline runtime/resetting_counter.go:44]: This is the only difference, compared to the `StandardCounter`.

File: runtime/resetting_histogram.go [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+	}
+}
+
+// Update is a no-op.
+func (NilResettingHistogram) Update(int64) {}
+
+// Clear is a no-op.
+func (NilResettingHistogram) Clear() {}
+
+func (NilResettingHistogram) Count() int64 {
```
Comment (coryschwartz) [inline runtime/resetting_histogram.go]: Why panic and not just a zero-value? If it returns a zero-value, NilResettingHistogram can be used in place of a regular Histogram. If it panics, it cannot.

Comment (coryschwartz) [reply runtime/resetting_histogram.go]: Nil values is what go-metrics-uses on their NilHistogram 

https://github.com/rcrowley/go-metrics/blob/10cdbea86bc0b922507ce91d43f4f78a5bea2e8f/histogram.go#L133

File: runtime/resetting_histogram.go [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+}
+
+// Snapshot resets the timer and returns a read-only copy of its sorted contents.
+func (t *StandardResettingHistogram) Snapshot() Histogram {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	currentValues := t.values
+	t.values = make([]int64, 0, InitialResettingHistogramSliceCap)
+
+	sort.Sort(Int64Slice(currentValues))
```
Comment (coryschwartz) [inline runtime/resetting_histogram.go]: We could avoid the single-purpose type below.

```suggestion
	sort.Slice(currentValues, func(i, j int) bool { currentValues[i] < currentValues[j] })
```

File: runtime/resetting_histogram.go [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+// Mean returns the mean of the snapshotted values
+func (t *ResettingHistogramSnapshot) Mean() float64 {
+	if !t.calculated {
+		t.calc([]float64{})
+	}
+
+	return t.mean
+}
+
+func (t *ResettingHistogramSnapshot) calc(percentiles []float64) {
```
Comment (coryschwartz) [inline runtime/resetting_histogram.go]: What stops this from being a regular go-metrics snapshot?

File: runtime/resetting_histogram.go (hunk 6 above)
Comment (nonsense) [reply runtime/resetting_histogram.go]: The idea is that you `panic` when the type is used incorrectly, for example see `HistogramSnapshot` on `go-metrics`. The idea here is the same - you are not supposed to call `Count` on a `Histogram`, but only on a `Snapshot`.

Same for all the other calls.

---

I don't know why `NilHistogram` is returning `nil` values for some of these methods, this seems inconsistent with the rest of the calls in the library. Some of these methods should be called only on a `Snapshot`.

File: runtime/resetting_histogram.go (hunk 8 above)
Comment (nonsense) [reply runtime/resetting_histogram.go]: Not sure what you mean with that. The `calc` function is different, and the backend data structure is different, so we can't use the `HistogramSnapshot`.

Maybe it'd be possible to push this to the `Sample` type, but this is not really a sample, and we are only moving the logic one-level further.

File: runtime/resetting_histogram.go (hunk 6 above)
Comment (nonsense) [reply runtime/resetting_histogram.go]: Ultimately for `NilResettingHistogram`, I don't really care if we panic or return nil values, but for the type used to collect the measurements, we should make sure that the user is using it correctly and we panic if it is not. Then once a concrete type is changed with the `NilResettingHistogram` (when metrics are disabled), it doesn't really matter.

The panics in `NilResettingHistogram` should correspond to the panics in `StandardResettingHistogram`. A user is not supposed to call anything but `Update`, `Snapshot`, `Clear` on `StandardResettingHistogram`.

File: runtime/resetting_histogram.go (hunk 7 above)
Comment (nonsense) [reply runtime/resetting_histogram.go]: I agree, this was legacy pre Go1.8 code, your suggestion is better now!

File: runtime/metrics.go (hunk 4 above)
Comment (raulk) [reply runtime/metrics.go]: Speculation ;-)

From the perspective of InfluxDB, the data says otherwise: https://www.influxdata.com/blog/influxdb-outperforms-graphite-in-time-series-data-metrics-benchmark/ (admittedly it is a marketing post, but it has stood the test of time; wasn't able to find disputes or criticism online).

In a 10k node scenario with 5 seconds frequency, we're looking at 2k concurrent batches -- should be fine.

Or were you concerned about the cost on the client-side?

File: runtime/resetting_counter.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+import (
+	"sync/atomic"
+
+	"github.com/rcrowley/go-metrics"
+)
+
+// GetOrRegisterCounter returns an existing Counter or constructs and registers
+// a new StandardCounter.
+func GetOrRegisterResettingCounter(name string, r metrics.Registry) metrics.Counter {
```
Comment (raulk) [inline runtime/resetting_counter.go]: This constructor is not needed, as far as I can tell. It will pollute the godocs for no reason.

File: runtime/resetting_counter.go [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+	if nil == r {
+		r = metrics.DefaultRegistry
+	}
+	return r.GetOrRegister(name, NewResettingCounter).(Counter)
+}
+
+// NewResettingCounter constructs a new StandardCounter.
+func NewResettingCounter() Counter {
```
Comment (raulk) [inline runtime/resetting_counter.go]: This can be made private, along with the type itself. No reason to pollute the godocs.

File: runtime/resetting_histogram.go [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+	"sync"
+
+	"github.com/rcrowley/go-metrics"
+)
+
+// Initial slice capacity for the values stored in a ResettingHistogram
+const InitialResettingHistogramSliceCap = 10
+
+// NewResettingHistogram constructs a new StandardResettingHistogram
+func NewResettingHistogram() Histogram {
```
Comment (raulk) [inline runtime/resetting_histogram.go]: Same comments as above. Let's make all these constructors private. The user should be constructing these through the runenv, and having these as public members muddies the docs.

File: runtime/resetting_histogram.go (hunk 6 above)
Comment (raulk) [reply runtime/resetting_histogram.go]: Yeah, let's not panic here. Let's return zero values. If we think that the component that iterates over the registry and populates the internal data model (that gets flushed to a file and to influx) should skip over these entirely (i.e. simply not process them, rather than sending 0 values), then we should do something else, such as adding a `Omitter` interface:

```go
type Omitter interface {
    IsOmitted() bool
}
```

That component would type assert against this interface and decide what to do. But it's not worth the pain. Let's just return zero values.

File: runtime/resetting_histogram.go [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+}
+
+// StandardResettingHistogram is used for storing aggregated values for timers, which are reset on every flush interval.
+type StandardResettingHistogram struct {
+	values []int64
+	mutex  sync.Mutex
+}
+
+func (t *StandardResettingHistogram) Count() int64 {
+	panic("Count called on a StandardResettingHistogram")
```
Comment (raulk) [inline runtime/resetting_histogram.go]: I'm not sure I follow why these methods panic? They should be functionally equivalent to `StandardHistogram`: https://github.com/rcrowley/go-metrics/blob/10cdbea86bc0b922507ce91d43f4f78a5bea2e8f/histogram.go#L162.

File: runtime/resetting_histogram.go [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+func (t *StandardResettingHistogram) Values() []int64 {
+	return t.values
+}
+
+// Snapshot resets the timer and returns a read-only copy of its sorted contents.
+func (t *StandardResettingHistogram) Snapshot() Histogram {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	currentValues := t.values
+	t.values = make([]int64, 0, InitialResettingHistogramSliceCap)
```
Comment (raulk) [inline runtime/resetting_histogram.go:141]: This is kinda terrible for GC; we should be using an `[]int64` pool, and returning the slice to the pool either via explicit refcount/close or through a finalizer.

File: runtime/resetting_histogram.go [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+
+	return &ResettingHistogramSnapshot{
+		values: currentValues,
+	}
+}
+
+func (t *StandardResettingHistogram) Clear() {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	t.values = make([]int64, 0, InitialResettingHistogramSliceCap)
```
Comment (raulk) [inline runtime/resetting_histogram.go]: Let's reslice to a 0-length slice instead: `t.values = t.values[:0]`

File: runtime/resetting_histogram.go [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+
+// Record the duration of an event.
+func (t *StandardResettingHistogram) Update(d int64) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	t.values = append(t.values, d)
+}
+
+// ResettingHistogramSnapshot is a point-in-time copy of another ResettingHistogram.
+type ResettingHistogramSnapshot struct {
```
Comment (raulk) [inline runtime/resetting_histogram.go]: Can/should be a private type.

File: runtime/resetting_histogram.go [hunk 16]
```diff
@@ -0,0 +1,298 @@
...
+
+			t.thresholdBoundaries[i] = float64(thresholdBoundary)
+		}
+
+		sum := cumulativeValues[count-1]
+		t.mean = float64(sum) / float64(count)
+	} else {
+		t.thresholdBoundaries = make([]float64, len(percentiles))
+		t.mean = 0
+	}
```
Comment (raulk) [inline runtime/resetting_histogram.go]: Easier to read if you short-circuit the zero case at the top and return immediately.

File: runtime/resetting_histogram.go [hunk 17]
```diff
@@ -0,0 +1,298 @@
...
+// Percentiles returns the boundaries for the input percentiles.
+func (t *ResettingHistogramSnapshot) Percentiles(percentiles []float64) []float64 {
+	t.calc(percentiles)
+
+	return t.thresholdBoundaries
+}
+
+// Mean returns the mean of the snapshotted values
+func (t *ResettingHistogramSnapshot) Mean() float64 {
+	if !t.calculated {
```
Comment (raulk) [inline runtime/resetting_histogram.go:279]: This is racy and rather weird, because the mean and the state is over-written on every re-calculation, and the input for the recalculation can change. I suggest you only store the mean in the state, because it's the only value that gets re-used from multiple getters. 

The threshold boundaries should not be part of the state because they don't get reused.

**Review by raulk — CHANGES_REQUESTED** [review]

File: runtime/resetting_histogram.go (hunk 12 above)
Comment (nonsense) [reply runtime/resetting_histogram.go]: It would be terrible to recalculate on every call to Min, Max, Variance, etc. if you call them directly on the Histogram. The idea is to call a Snapshot every N-seconds (interval), and calculate once.

The panics signal to the user that if they call Histogram.Min they are doing it wrong.

File: runtime/resetting_histogram.go (hunk 14 above)
Comment (nonsense) [reply runtime/resetting_histogram.go]: Isn't this going to overwrite the `Snapshot` values... you can concurrently hold a `Snapshot` and continue to use and update your `Histogram`.

File: runtime/resetting_histogram.go (hunk 13 above)
Comment (nonsense) [reply runtime/resetting_histogram.go:141]: `t.values` is transferred to the `Snapshot`, and we initialise a new slice for the `Histogram`. If you are suggesting to re-use the memory of `t.values` for the next values in the Histogram, we have to copy them over for the Snapshot, which is essentially the same thing.

File: runtime/resetting_histogram.go (hunk 17 above)
Comment (nonsense) [reply runtime/resetting_histogram.go:279]: Fixed. Generally you have one actor interacting with a snapshot, but I agree that better to make this non-racy.

File: runtime/resetting_histogram.go (hunk 6 above)
Comment (nonsense) [reply runtime/resetting_histogram.go]: Returning zero values now. The component that iterates over these, should be calling `Snapshot` and working on the snapshot, and not on the `Histogram` itself.

For what is worth, you have to have your values sorted to return Min, Max, etc. and it doesn't make sense to sort values multiple times. If a user is calling Min or Max on the Histogram directly, they are doing it wrong.

**Comment by nonsense** [comment]
> Speculation ;-)

You are right, I don't mind reverting to 5sec, I just think we are wasting space, because the 10sec. granularity has appeared to work well for me for now ;)

File: runtime/metrics.go (hunk 4 above)
Comment (nonsense) [reply runtime/metrics.go]: You are right, I don't mind reverting to 5sec, I just think we are wasting disk space, because the 10sec. granularity has appeared to work well for me for now ;)

100 measurements * 10sec. * 10000 node testplans is a lot of data on disk, specially if we want to do longer retention in the future.

Comment (raulk) [reply runtime/metrics.go]: Sure, but we shouldn't assume 10k is the _default_ size of tests. Users can easily change the frequency by calling `runenv.D().SetFrequency()` for larger tests. Let's let the user take the tradeoff they want.

File: runtime/resetting_histogram.go (hunk 13 above)
Comment (raulk) [reply runtime/resetting_histogram.go:141]: I should've clarified. We could set a finalizer **on the `Snapshot`** that returns the slice to the pool of this histogram. Basically I'm thinking that the pattern will be:

* Long-lived resetting histogram.
* Every 5 seconds, call `Snapshot()` => transfers `t.values` onto the Snapshot and makes a new slice.
* The `Snapshot` gets used, probably immediately and just once, and then it goes out of scope.
  * If we add a `Close()` method that returns the slice (reset to 0 length) to the histogram, so that it's available to be re-used the next time we rotate, we will effectively used as many slices as Snapshots are alive/retained (probably just 1 at a time) + the active slice, max. 2 or 3 slices.
  * With a finalizer, things aren't deterministic, but still we can curtail allocs by some amount.

With the current code, we will just allocate and GC naïvely.

Maybe this is premature optimisation, but wanted to raise this as I think it might bite us in the future.

File: runtime/resetting_histogram.go (hunk 14 above)
Comment (raulk) [reply runtime/resetting_histogram.go]: This is happening on the histogram, no? The `Snapshot` is not referencing `t.values`.

File: runtime/resetting_histogram.go (hunk 12 above)
Comment (raulk) [reply runtime/resetting_histogram.go]: Yeah, I agree, but the upstream library does support this calculation and I think we should do too. I don't see any reason to drift?

File: runtime/resetting_histogram.go (hunk 13 above)
Comment (nonsense) [reply runtime/resetting_histogram.go:141]: I see your point now. I think allocating a bit of memory (specially when we don't have thousands of histograms in our plans) a few times every 5-10sec. is fine.

We can totally do a memory pool, but feels like premature optimisation.

I suspect that when we go to TaaS and we start keeping longer retention for metrics, we might downgrade the intervals to 10sec. or even 30sec. intervals, which should reduce memory allocations even further.

File: runtime/resetting_histogram.go (hunk 12 above)
Comment (nonsense) [reply runtime/resetting_histogram.go]: The upstream library is operating on samples, this data structure is holding all values and mostly calculating percentiles on them.

If we return these calculations on the `StandardResettingHistogram`, then what is the point of having a `Snapshot`? Also we'd be locking the `Histogram`, and blocking any new updates, while processing the current values.

I really don't see a point in supporting these function on the `Histogram` level.

Comment (nonsense) [reply runtime/resetting_histogram.go]: Also you just asked me to unexport all these types - nobody, but our `runtime` package code can actually use these. We are not building a library that is extending the `go-metrics` apparently, that other folks can use. Our code is using these data structures correctly, by generating a snapshot on every interval prior to emitting the measurements.

**Comment by nonsense** [comment]
Failed tests seem unrelated.

File: runtime/resetting_histogram.go (hunk 13 above)
Comment (raulk) [reply runtime/resetting_histogram.go:141]: We can leave it like this for now. I opened an issue to track: https://github.com/testground/testground/issues/1021.

File: runtime/metrics_types.go (hunk 2 above)
Comment (raulk) [reply runtime/metrics_types.go]: I don't think it hurts to mimic the behaviour we're replacing. There's value in calculating these derivatives here -- as calculating them on Influx is would already be doing so with a lesser resolution.

Comment (nonsense) [reply runtime/metrics_types.go]: 99% of charts won't be graphing those. My thinking is that we include them, if someone really needs them or finds them valuable. Basicaly I meant YAGNI :D

Comment (raulk) [reply runtime/metrics_types.go]: What's the technical reason to drop them? Are they costly to calculate?

File: runtime/metrics_types.go [hunk 18]
```diff
@@ -151,14 +149,13 @@ func NewMetric(name string, i interface{}) *Metric {
...
 		m.Measures["p999"] = p[4]
-		m.Measures["p9999"] = p[5]
 		m.Measures["m1"] = s.Rate1()
 		m.Measures["m5"] = s.Rate5()
 		m.Measures["m15"] = s.Rate15()
 		m.Measures["meanrate"] = s.RateMean()
 
 	default:
-		panic("unexpected metric type")
+		panic(fmt.Sprintf("unexpected metric type: %v", reflect.TypeOf(v)))
```
Comment (raulk) [inline runtime/metrics_types.go:162]: Cool.

**Review by raulk — APPROVED** [review]
LGTM. I just pushed a commit in f709bf3 to not expose internal type names in panic messages.

File: runtime/metrics_types.go (hunk 2 above)
Comment (nonsense) [reply runtime/metrics_types.go]: One of the main use-cases to include this type is to measure latencies. We care about mean, 95%, 99% and max, when graphing those. We don't want to miss spikes and measurements that happen in interval N to affect measurements in the next interval N+1, hence the resetting.

It is not clear to me why someone would want to graph `stddev` and `variance` for latencies at all - also one is a derivation of the other. Basically we are adding 2 fields that probably won't be used as part of every measurement, sending them across the wire and persisting them - it seems wasteful to me.

Folks are welcome to calculate these off-band with other tools we have build in the SDK, for example the outputs.

---

Why add more data to every measurement, when we don't have an immediate use-case for it?

---

That's my thinking.

---

Bottom line: If someone comes with a use-case that needs those two - it is trivial to add them later... I don't see a reason to add them as part of the SDK today.

File: runtime/resetting_histogram.go (hunk 8 above)
Comment (raulk) [reply runtime/resetting_histogram.go]: @coryschwartz we can't operate a go-metrics `HistogramSnapshot` because its state is not public: https://pkg.go.dev/github.com/rcrowley/go-metrics?tab=doc#HistogramSnapshot.

**Comment by raulk** [comment]
@nonsense merge when ready. We just have this thread, where I don't understand the technical reasons for dropping something that's fairly easy to calculate: https://github.com/testground/sdk-go/pull/14#discussion_r424400364

File: runtime/metrics_types.go (hunk 18 above)
Comment (nonsense) [reply runtime/metrics_types.go:162]: I got the panic once, so I think its helpful :)

**Comment by nonsense** [comment]
> @nonsense merge when ready. We just have this thread, where I don't understand the technical reasons for dropping something that's fairly easy to calculate: #14 (comment)

Reverted all fields as discussed.
